[SYCL][Docs] Update sycl_ext_intel_usm_address_spaces and fix ctors (#7680)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_intel_usm_address_spaces.asciidoc

specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_INTEL_USM_ADDRESS_SPACES` to one of the values defined in the table
below.  Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

== Modifications to SYCL 2020

The following sections contain the related changes and additions to the SYCL
2020 specification relating to this extension.

=== SYCL Device Memory Model

Add to the end of the definition of global memory:
Global memory is a virtual address space which overlaps the device and host
address spaces.

Add two new memory regions as follows:

*Device memory* is a sub-region of global memory that is not directly accessible
by the host.  Global accessors and USM allocations of the device alloc type
reside in this address space.

> *Host memory* is a sub-region of global memory.  USM pointers allocated with the
> host alloc type reside in this address space.

Reviewer: ```suggestion
*Host memory* is a sub-region of global memory.  USM allocations whose kind is `usm::alloc::host` reside in this address space.
```

Use the same wording here as above for "device memory".

also allow `access::address_space::ext_intel_global_device_space` as follows:

--
[options="header"]
|===
| Constructor | Description
a|
```c++
template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder>
multi_ptr(
    accessor<ElementType, Dimensions, Mode, target::device, IsPlaceholder>);
```
| Available only when:
`Space == access::address_space::global_space \|\| Space == access::address_space::ext_intel_global_device_space \|\| Space == access::address_space::generic_space`.

Constructs a `multi_ptr` from an accessor of `target::device`.

This constructor may only be called from within a command.
|===
--

=== Explicit Pointer Aliases

Add `device_ptr` and `host_ptr` aliases to the list of `multi_ptr` aliases as
follows:
```c++
template<typename ElementType, access::decorated IsDecorated = access::decorated::legacy>
using device_ptr = multi_ptr<ElementType, access::address_space::ext_intel_global_device_space, IsDecorated>

template<typename ElementType, access::decorated IsDecorated = access::decorated::legacy>
using host_ptr = multi_ptr<ElementType, access::address_space::ext_intel_global_host_space, IsDecorated>

> ```

Reviewer: Clarify which namespace these are in.  I think they are currently in `sycl`, but that's not the right place.  They should either be in `sycl::ext::intel` or `sycl::ext::intel::experimental`, depending on the status of this extension.

I think we should deprecate the existing aliases and add new ones in the correct namespace.

Author: Currently it is in the `sycl` namespace, but I agree they should be moved. Should we also add `raw_` and `decorated_` variants while we are in here? @GarveyJoe - Are you okay if we just move them or would you prefer we keep the old ones in the `sycl` namespace as deprecated for now?

Reviewer 2: Please keep the old ones working with deprecation warnings for now so that we can migrate gradually.  FYI, @ajaykumarkannan
